service/collector/usage.go

```diff
@@ -0,0 +1,178 @@
+			"name",
+		},
+		nil,
+	)
+	usageLimitDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName("azure_operator", "usage", "limit"),
+		"Usage limit of specific Quotas as defined by Azure.",
+		[]string{
+			"name",
+		},
+		nil,
+	)
+)
+
+type UsageConfig struct {
+	G8sClient versioned.Interface
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+
+	// EnvironmentName is the name of the Azure environment used to compute the
+	// azure.Environment type. See also
+	// https://godoc.org/github.com/Azure/go-autorest/autorest/azure#Environment.
+	EnvironmentName string
+	Location        string
+}
+
+type Usage struct {
+	g8sClient versioned.Interface
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
+
+	environmentName string
+	location        string
+}
+
+func NewUsage(config UsageConfig) (*Usage, error) {
+	if config.G8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.G8sClient must not be empty", config)
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.EnvironmentName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.EnvironmentName must not be empty", config)
+	}
+	if config.Location == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Location must not be empty", config)
+	}
+
+	u := &Usage{
+		g8sClient: config.G8sClient,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+
+		environmentName: config.EnvironmentName,
+		location:        config.Location,
+	}
+
+	return u, nil
+}
+
+func (u *Usage) Collect(ch chan<- prometheus.Metric) error {
+	// We need all CRs to gather all subscriptions below.
+	var crs []providerv1alpha1.AzureConfig
+	{
+		mark := ""
+		page := 0
+		for page == 0 || len(mark) > 0 {
+			opts := metav1.ListOptions{
+				Continue: mark,
+			}
+			list, err := u.g8sClient.ProviderV1alpha1().AzureConfigs("").List(opts)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			crs = append(crs, list.Items...)
+
+			mark = list.Continue
+			page++
+		}
+	}
+
+	// We generate clients and group them by subscription.
+	clients := map[string]*compute.UsageClient{}
+	{
+		for _, cr := range crs {
+			c, err := u.getUsageClient(cr)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			clients[c.SubscriptionID] = c
+		}
+	}
+
+	// We track usage metrics for each client, which is unique per subscription.
+	// That way we prevent duplicated metrics.
+	for _, c := range clients {
+		r, err := c.List(context.Background(), u.location)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for r.NotDone() {
+			for _, v := range r.Values() {
+				ch <- prometheus.MustNewConstMetric(
+					usageCurrentDesc,
+					prometheus.GaugeValue,
+					float64(*v.CurrentValue),
+					*v.Name.LocalizedValue,
```

Comment:
What does LocalizedValue mean here? Does this mean that if the subscription is german we get the german name here?

Reply:
There are two "values". One is camel cased and has sometimes insufficient information as I found out. The localized thingy is what you see in Prometheus. It is more human readable. I thought this is the better option. 

<img width="1920" alt="Screenshot 2019-07-04 at 14 08 45" src="https://user-images.githubusercontent.com/552769/60666161-af839600-9e66-11e9-8cbc-0020d4a51d9f.png">
<img width="1920" alt="Screenshot 2019-07-04 at 14 08 26" src="https://user-images.githubusercontent.com/552769/60666162-af839600-9e66-11e9-839d-eef1e7a089f3.png">

Reply:
I see, makes sense.

I am a little concerned about adding 53 new series to prometheus and what impact on memory usage it will have. But we can try and see.

Reply:
It is 106 per subscription, but why should that matter? That is really not that much data IMO.

Reply:
I remember that we limitied the amount of series we get from tenants multiple times to improve performance of prometheus. I do not have a reference point as to what is `a lot` of data unfortunately - hence my concern.

You are right that it should be fine but I am lacking a sense of scale here.

Reply:
I also would prefer to have only metrics that we care for. But there are places where we can probably shave way more.

As for the localized name won't it be awkward to use that for creating alerting rules? I wouldn't be surprised if those names change over time.

Reply:
What I found weird with the other names is that the CPU information or VM family wasn't always super clear. So far I did not write an alert but when we do it is rather easy to improve now after most of the hustle got sorted out.

Reply:
Let's iterate.